### PR TITLE
Support vmapped compiled autograd.Functions

### LINKED ISSUE #129845: vmap + autograd.Function + torch.compile don't work together
### 🚀 The feature, motivation and pitch

```
from torch.func import vmap

class Foo(torch.autograd.Function):
    generate_vmap_rule = True

    @staticmethod
    def forward(i):
        return torch.tanh(i) * 2

    @staticmethod
    def setup_context(ctx, inputs, output):
        pass

    @staticmethod
    def backward(ctx, i):
        return i

def f(x):
    return x.tanh()

f = vmap(f)
f(torch.randn(10, requires_grad=True))
torch.compile(f, fullgraph=True)(torch.randn(10, requires_grad=True))
```

cc: @zou3519 @oulgen 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186366

Dynamo lowers captured autograd.Function calls through an internal
ApplyTemplate autograd.Function. Two pieces were missing for the issue's
composition of vmap, autograd.Function with generate_vmap_rule=True, and
compile.

First, ctx.mark_non_differentiable handling assumed that forward returned a
list or tuple, so a single tensor output tripped an assertion before the HOP
could be emitted. This now mirrors mark_dirty and maps a marked single tensor
output to output index 0.

Second, ApplyTemplate already has setup_context but did not opt into generated
vmap rules. As a result, functorch rejected vmap over the reconstructed compiled
autograd.Function even when the original Function used generated vmap support.
Setting generate_vmap_rule=True on ApplyTemplate lets the existing generated
vmap machinery run the traced forward/backward bodies under vmap.

The alternative would be to special-case flex attention or leave the old test
commented out, but that would preserve the broken autograd.Function composition
instead of fixing the shared capture path.

Fixes #129845
Generated by my agent

Test Plan:
- pytest test/dynamo/test_autograd_function.py -k 'test_mark_single_output_non_differentiable or test_new_style_autograd_function_with_vmap_compiled'
- pytest test/inductor/test_flex_attention.py -k test_autograd_function_in_score_mod
- pytest test/dynamo/test_autograd_function.py
- lintrunner -a
- git diff --check

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the dynamo_wrapped failures were stale expected-failure markers: the PR now makes those mark_non_differentiable tests pass under Dynamo, so I removed the obsolete expected-failure files. I also tightened the vmap fix to preserve the original autograd.Function vmap contract and removed the unrelated ciflow/torchtitan label.

**Comment by jansel** [comment]
@claude review these changes

File: torch/_functorch/autograd_function.py [hunk 1]
```diff
@@ -819,10 +819,50 @@ def __call__(
...
+        current_functorch_interpreter = torch._C._functorch.peek_interpreter_stack()
+        is_under_vmap = (
+            current_functorch_interpreter is not None
+            and current_functorch_interpreter.key()
+            is torch._C._functorch.TransformType.Vmap
+        ) or any(
+            torch._C._functorch.is_batchedtensor(arg)
+            for arg in fwd_args
+            if isinstance(arg, torch.Tensor)
+        )
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_functorch/autograd_function.py:835]: My agent says I changed this vmap detection to compare TransformType.Vmap with == in torch/_functorch/autograd_function.py, so it no longer relies on pybind enum object identity.

File: torch/_dynamo/graph_break_registry.json [hunk 2]
```diff
@@ -4938,6 +4938,30 @@
...
+  "GB7811": [
+    {
+      "Gb_type": "Unsupported ambiguous autograd.Function vmap",
+      "Context": "call_apply {self} {args} {kwargs}",
+      "Explanation": "Dynamo does not support tracing `torch.autograd.Function` subclasses that set `generate_vmap_rule=True` and define a custom `vmap` staticmethod.",
+      "Hints": [
+        "Set `generate_vmap_rule=False` or remove the custom ",
+        "`vmap` staticmethod.",
+        "It may be possible to write Dynamo tracing rules for this code. Please report an issue to PyTorch if you encounter this graph break often and it is causing performance issues."
+      ]
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_dynamo/graph_break_registry.json:5067]: My agent says I combined this graph-break hint into one string in torch/_dynamo/variables/misc.py and the synced torch/_dynamo/graph_break_registry.json entry, so the rendered hint is no longer split.

File: torch/_dynamo/variables/misc.py [hunk 3]
```diff
@@ -972,6 +972,52 @@ def visit(vt: VariableTracker) -> None:
...
+                example_value = vt.as_proxy().node.meta["example_value"]
+                if torch._C._functorch.is_batchedtensor(example_value):
+                    has_batched_tensor_arg = True
+
+        VariableTracker.visit(visit_batched_tensor_arg, (args, kwargs))
+        is_under_vmap = (
+            current_functorch_interpreter is not None
+            and current_functorch_interpreter.key()
+            is torch._C._functorch.TransformType.Vmap
+        ) or has_batched_tensor_arg
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_dynamo/variables/misc.py:1003]: My agent says I changed the matching Dynamo vmap detection to compare TransformType.Vmap with == in torch/_dynamo/variables/misc.py, so the check is by enum value.

File: torch/_dynamo/graph_break_registry.json [hunk 4]
```diff
@@ -4938,6 +4938,30 @@
...
+  "GB3449": [
+    {
+      "Gb_type": "Unsupported autograd.Function vmap",
+      "Context": "call_apply {self} {args} {kwargs}",
+      "Explanation": "Dynamo only supports tracing `torch.autograd.Function` under vmap when `generate_vmap_rule=True`.",
+      "Hints": [
+        "Set `generate_vmap_rule=True` if generated vmap ",
+        "rules are valid for this autograd.Function.",
+        "It may be possible to write Dynamo tracing rules for this code. Please report an issue to PyTorch if you encounter this graph break often and it is causing performance issues."
+      ]
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_dynamo/graph_break_registry.json:5078]: My agent says I combined this graph-break hint into one string in torch/_dynamo/variables/misc.py and the synced torch/_dynamo/graph_break_registry.json entry as well.

File: torch/_dynamo/variables/misc.py [hunk 5]
```diff
@@ -984,6 +984,50 @@ def visit(vt: VariableTracker) -> None:
...
+                    explanation="Dynamo does not support tracing "
+                    "`torch.autograd.Function` subclasses that set "
+                    "`generate_vmap_rule=True` and define a custom `vmap` "
+                    "staticmethod.",
+                    hints=[
+                        "Set `generate_vmap_rule=False` or remove the custom `vmap` staticmethod.",
+                        *graph_break_hints.SUPPORTABLE,
+                    ],
+                )
+            if not self.fn_cls.generate_vmap_rule:
```
Comment (aorenste) [inline torch/_dynamo/variables/misc.py:1018]: [cleanup] This `not generate_vmap_rule` branch also swallows the case where the Function defines a custom `vmap` staticmethod (`has_custom_vmap=True`, `generate_vmap_rule=False`) - a Function that actually *has* working eager vmap support. Those users are told to "Set `generate_vmap_rule=True`", which discards their custom rule and is not necessarily an equivalent substitute. The runtime check in `torch/_functorch/autograd_function.py` already distinguishes this with a custom-vmap-specific message ("...subclasses that define a custom vmap staticmethod"); branching on `has_custom_vmap` here before the generic message would make the trace-time and runtime layers consistent. Low priority. Note `test_autograd_function_custom_vmap_compiled_unsupported` asserts on the `generate_vmap_rule=True` substring, so it would need updating if the message changes.

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

[blocker] The runtime vmap-enforcement block added to `AutogradFunctionApply.__call__` in `torch/_functorch/autograd_function.py` (the `is_under_vmap` check that raises `RuntimeError`) is new, reachable, user-facing behavior with no test. The two trace-time tests (`test_autograd_function_without_vmap_support_compiled`, `test_autograd_function_custom_vmap_compiled_unsupported`) exercise the `misc.py` path instead: a plain `torch.compile`d function later called under `torch.vmap` recompiles (the batched-tensor input fails the dispatch-key guard) and graph-breaks before the HOP is ever executed under vmap, so the runtime raise is never hit. That raise is only reachable when a graph captured *without* vmap is replayed *under* vmap (export / captured-graph reuse), which is reproducible. Please add a regression test for that path before merge - capture or export a graph containing the `autograd_function_apply` HOP, then run it under `torch.vmap` and assert the `RuntimeError` - otherwise the new error path ships unverified and can regress silently.

**Comment by jansel** [comment]
My agent says I addressed the requested runtime coverage by adding `test_autograd_function_without_vmap_support_replayed_under_vmap`. The test captures a graph containing `autograd_function_apply`, replays that recorded graph under `torch.vmap`, and asserts the `ApplyTemplate` runtime `RuntimeError`, so it does not pass through Dynamo’s trace-time `Unsupported` path. I also split the custom-vmap trace-time graph break into its own message and re-ran the focused Dynamo tests, the full `test/dynamo/test_autograd_function.py`, graph-break registry checks, `lintrunner -a`, and `git diff --check`.

File: torch/_dynamo/variables/misc.py (hunk 5 above)
Comment (jansel) [reply torch/_dynamo/variables/misc.py:1018]: My agent says I split the custom-vmap trace-time case into its own `Unsupported custom autograd.Function vmap` graph break and updated the test assertion, so users with a custom `vmap` rule no longer get told to switch to generated vmap rules.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
